src/librustc_trans/trans/base.rs

```diff
@@ -722,13 +736,20 @@ pub fn invoke<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
         let normal_bcx = bcx.fcx.new_temp_block("normal-return");
         let landing_pad = bcx.fcx.get_landing_pad();
 
-        let llresult = Invoke(bcx,
+        let mut llresult = Invoke(bcx,
                               llfn,
                               &llargs[..],
                               normal_bcx.llbb,
                               landing_pad,
                               Some(attributes),
                               debug_loc);
+        if let ty::FnConverging(ty) = ret_ty {
+            if return_type_is_void(bcx.ccx(), ty) {
+                llresult = C_nil(bcx.ccx());
+            }
+        } else {
+            llresult = C_nil(bcx.ccx());
+        }
```

Comment:
I'd prefer to add that check after the if-else block, so it can be merged with the same one in the else-arm.